pybamm/solvers/c_solvers/idaklu/Expressions/IREE/IREEFunctions.cpp

```diff
@@ -0,0 +1,303 @@
+#include "IREEFunctions.hpp"
+#include "iree_jit.hpp"
+#include <iostream>
+#include <fstream>
+#include <string>
+#include <vector>
+#include <regex>
+
+IREEFunction::IREEFunction(const BaseFunctionType &f) : Expression(), m_func(f)
+{
+  DEBUG("IreeFunction constructor");
+  const std::string& mlir = f.get_mlir();
+
+  // Parse IREE (MLIR) function string
+  if (mlir.size() == 0) {
+    DEBUG("Empty function --- skipping...");
+    return;
+  }
+
+  // Parse module name
+  std::regex module_name_regex("module @([^\\s]+)");  // Match until first whitespace
```

Comment:
does iree have a parser class that can you use to do this? (e.g. https://mlir.llvm.org/doxygen/classmlir_1_1detail_1_1Parser.html). This seems quite manual

Reply:
This is proving surprisingly difficult to implement. I can see two options:
1) working with IREE: the routines provided only seem to permit access to low-level memory details, using representations below the `tensor` syntax, which makes extraction difficult.
2) working with llvm-mlir: we have to compile-in the llvm/mlir libraries alongside iree; set-up new contexts and register dialects in order to parse the mlir string. This approach has potential but is so far proving non-trivial to get working in harmony with the iree solution we already have in place.

As this is turning into a bit of a blocker, I have decided to refactor the existing module parsing code out into it's own ModuleParser class that can be modified or wholesale replaced as needed at a later date.

Reply:
fair enough. Happy to go with the separate parsing class for now.